Transaction refactor

lukash opened the pull request:
Refactors the transaction package passing from the Goal onwards.

Untested on microdnf so far, likely to explode. I'm working on persistent reason storage (implementation mostly done but I left it out of this PR in case that will require more discussion) so that we can test microdnf through the ci stack.

Commit:
tests: Move the get_pkg() method to RepoFixture

Commit:
Move PackageSetIterator impl into .cpp

Commit:
Goal tests: Add test repo only where relevant

Commit:
libdnf/base/transaction: Clean up the Transaction(Item)Package class

Rename TransactionPackageItem to TransactionPackage, seems sufficient
(and shorter).

Use simple composition instead of multiple inheritance. Separating
transaction item definition into two classes is unnecessary, it can be a
single class. Inheriting from Package makes the get_reason() method
confusing, because the Package implementation returns something else
than the TransactionPackage implementation. This makes it more clear on
the interface which method is called.

Return std::optional<Package> (which is semantically more explicit than
a raw pointer) from get_replace() and rename to get_replaces().

Add docstrings, do minor cleanups.

Commit:
Pass TranasctionPackages from Goal and Transaction onwards

Cleans the TransactionPackage handling from Goal throughout to microdnf
and daemon transactions.

Commit:
tests: Add options to RepoFixture::get_pkg, improve error messages

Adds optional argument to query for an installed or available package
and an overload for querying from a specific repository (in case there
is more than one loaded and the package is in both).

In case the nevra would match more than one package, prints the packages
in test error output. In case no package matches, prints all packages in
the sack.

Commit:
libdnf-cli: Print an epmty line before Transaction Summary

Commit:
libdnf-cli: Add an ActionHeaderPrinter class

dmach commented on dnfdaemon-client/wrappers/dbus_transaction_package_wrapper.hpp:
```diff
@@ -0,0 +1,50 @@
+    DbusPackageWrapper get_package() const { return package; }
+
+private:
+    libdnf::transaction::TransactionItemAction action;
+    DbusPackageWrapper package;
```
`reason` might be needed too, because it has a different meaning than `package.get_reason()`. It's probably not necessary to add it in this PR.

lukash replied:
I already have it there, though just a static `UNKNOWN` for now, didn't go through adding it to dbus yet.

dmach commented on dnfdaemon-server/transaction.cpp:
```diff
@@ -0,0 +1,56 @@
+RpmTransactionItem::RpmTransactionItem(const libdnf::base::TransactionPackage & tspkg)
+    : TransactionItem(tspkg.get_package()) {
+    switch (tspkg.get_action()) {
+        case libdnf::transaction::TransactionItemAction::INSTALL:
+            action = Actions::INSTALL;
+            break;
+        case libdnf::transaction::TransactionItemAction::UPGRADE:
+            action = Actions::UPGRADE;
+            break;
+        case libdnf::transaction::TransactionItemAction::DOWNGRADE:
+            action = Actions::DOWNGRADE;
+            break;
+        case libdnf::transaction::TransactionItemAction::REINSTALL:
+            action = Actions::REINSTALL;
+            break;
+        case libdnf::transaction::TransactionItemAction::REMOVE:
+        case libdnf::transaction::TransactionItemAction::OBSOLETED:
+            action = Actions::ERASE;
+            break;
```
Just a note: I'd prefer removing `Actions` in favor of `libdnf::transaction::TransactionItemAction`.
But it's probably ok for now, we can discuss it in the next iteration(s).

dmach commented on include/libdnf/base/transaction_package.hpp:
```diff
@@ -0,0 +1,98 @@
+
+namespace libdnf::base {
+
+class TransactionPackage {
```
Just an idea: couldn't we use a template here?
I suppose variations of this object will be used with a different Package object.

Commit:
microdnf: Add a global try-catch for RuntimeError

Any exception inheriting from RuntimeError is an expected exception that
we need to catch and print the error, we don't want to crash on those.

dmach commented on include/libdnf/rpm/transaction.hpp:
```diff
@@ -274,6 +275,45 @@ class Transaction {
+        const std::vector<libdnf::base::TransactionPackage> & transaction_packages,
+        std::vector<std::unique_ptr<T>> & rpm_transaction_items_out) {
+        for (auto & tspkg : transaction_packages) {
+            auto fiddle_the_ptr = [&rpm_transaction_items_out](const libdnf::base::TransactionPackage & tspkg) {
```
Oh no, a lambda :)
Could you add a comment explaining what's going on?
Also, a more descriptive function would be nice.

lukash replied:
I'm not fond of the mechanism used, it was copy-pasted in each of the switch cases below, so I at least put it into the lambda. And tbh. I consider the name of the lambda quite fitting :nerd_face: I'm kinda hoping this will be improved (as the TODO in the method docstring says), but it may have to stay...

I'll add an explanatory comment.

dmach commented on libdnf-cli/output/transaction_table.hpp:
```diff
@@ -39,43 +39,140 @@ namespace libdnf::cli::output {
+    void print() {
+        std::cout << "Transaction Summary:\n";
+        if (installs != 0) {
+            std::cout << fmt::format(" {:15} {:4} packages\n", "Installing:", installs);
```
It might be worth replacing the code with libdnf::cli::output::KeyValueTable.

dmach reviewed (changes requested):
I know it's still a draft, but I managed to trigger an exception:
```
$ microdnf install iftop
$ microdnf remove iftop
libdnf::Exception: General libdnf exception: SQLite error on "/var/lib/dnf/history.sqlite": Reading a row failed: UNIQUE constraint failed: rpm.name, rpm.epoch, rpm.version, rpm.release, rpm.arch
```

lukash commented:
@dmach the crash is fixed, I'd leave the comments I didn't address for later iterations if you don't mind.

I've pushed a finalized version, which passes some basic CI stack tests I've enabled in this PR: https://github.com/rpm-software-management/ci-dnf-stack/pull/1013

dmach commented:
It's not perfect yet, but it's a step in the right direction.
Let's do couple more iterations in separate pull-requests.
